xcdat/averager.py

```diff
@@ -0,0 +1,458 @@
+      the variable domain
+    - Get axis weights for standard lat / lon domains
+    - Adjust weights to conform to (optional) regional boundary
+    - Compute spatial weighted average
+
+    Parameters
+    ----------
+    xvariable : xr.DataArray
+        xr.DataArray produced using ``open_variable``
+    lat_bounds : list, optional
+        List of upper and lower latitude boundaries [default None]
+    lon_bounds : list, optional
+        List of upper and lower longitude boundaries [default None]
+    axis : list, optional
+        List of dimensions to average over (default ['lat', 'lon'])
+    weight : str, optional
+        Weighting option to calculate weights proportional to area
+        ("weighted"), uniform weights ("unweighted"). Alternatively, a labelled
+        dataarray can be passed in as a custom set of weights.
+
+    Returns
+    -------
+    xr.DataArray
+        Dataarray that has been reduced via a weighted average.
+
+    Notes
+    -----
+    This assumes a certain pathology (the axis orientation simply does
+    not match). More complex issues may exist.
+
+    Examples
+    --------
+    Import:
+    >>> from xcdat.dataset import open_dataset
+    >>> from xcdat.variable import open_variable
+
+    Open dataset and dataarray:
+    >>> filename = 'test.nc'
+    >>> ds = open_dataset(filename)
+    >>> tas = open_variable(ds, 'tas')
+
+    Get global average time series:
+    >>> ts_global = spatial_average(tas)
+
+    Get time series in Nino 3.4 domain:
+    >>> ts_n34 = spatial_average(tas,
+                                 lat_bounds=[-5, 5],
+                                 lon_bounds=[-170, -120])
+
+    Get zonal mean time series:
+    >>> ts_zonal = spatial_average(tas, axis=['lon'])
+
+    Validate / fix bounds:
+    >>> validated_regional_bounds = validate_region_lon_bounds(tas.bounds.lon,
+                                                               regional_bounds)
+
+    Raises
+    ------
+    ValueError
+        Thrown if there is a conflict between specified bounds and the
+        dataarray bounds
+    """
+    # Check to see if the region bounds are within the domain of the dataset
+    if lon_bounds:
+        # if not, guess the axis orientation (i.e., -180 to 180 or 0 to 360) of
+        # the dataarray axes and then convert the region bounds to that
+        # orientation
+        if not _validate_region_lon_bounds(xvariable.bounds.lon, lon_bounds):
+            if np.max(xvariable.lon) <= 180:
+                lon_bounds = _switch_lon_bounds(lon_bounds, totype='to180')
+            else:
+                lon_bounds = _switch_lon_bounds(lon_bounds, totype='to360')
+            # after region bounds are adjusted, check to ensure they are
+            # consistent with the dataarray axes
+            if not _validate_region_lon_bounds(xvariable.bounds.lon,
+                                               lon_bounds):
+                # REVIEW: Should we throw a more specific error?
```

Comment:
ValueError is correct. We can create our own custom xcdat errors if needed, but low priority.